src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/ReadYamlStep.java

```diff
@@ -0,0 +1,184 @@
+import org.yaml.snakeyaml.Yaml;
+
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.TaskListener;
+
+/**
+ * Reads a yaml file from the workspace.
+ *
+ * @author Philippe GRANET &lt;philippe.granet@gmail.com&gt;.
+ */
+public class ReadYamlStep extends AbstractStepImpl {
+
+	private String file;
+	private String text;
+
+	@DataBoundConstructor
+	public ReadYamlStep() {
+	}
+
+	/**
+	 * Name of the yaml file to read.
+	 *
+	 * @return file name
+	 */
+	public String getFile() {
+		return file;
+	}
+
+	/**
+	 * Name of the yaml file to read.
+	 *
+	 * @param file
+	 *            file name
+	 */
+	@DataBoundSetter
+	public void setFile(String file) {
+		this.file = file;
+	}
+
+	/**
+	 * A String containing properties formatted data.
+	 *
+	 * @return text to parse
+	 */
+	public String getText() {
+		return text;
+	}
+
+	/**
+	 * A String containing properties formatted data.
+	 *
+	 * @param text
+	 *            text to parse
+	 */
+	@DataBoundSetter
+	public void setText(String text) {
+		this.text = text;
+	}
+
+	@Extension
+	public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+		public DescriptorImpl() {
+			super(Execution.class);
+		}
+
+		@Override
+		public String getFunctionName() {
+			return "readYaml";
+		}
+
+		@Override
+		public String getDisplayName() {
+			return "Read yaml from files in the workspace or text.";
+		}
+
+		@Override
+		public Step newInstance(Map<String, Object> arguments) throws Exception {
+			ReadYamlStep step = new ReadYamlStep();
+			if (arguments.containsKey("file")) {
+				Object file = arguments.get("file");
+				if (file != null) {
+					step.setFile(file.toString());
+				}
+			}
+			if (arguments.containsKey("text")) {
+				Object text = arguments.get("text");
+				if (text != null) {
+					step.setText(text.toString());
+				}
+			}
+			if (isBlank(step.getFile()) && isBlank(step.getText())) {
+				throw new IllegalArgumentException("At least one of file or text needs to be provided to readYaml.");
+			}
+			return step;
+		}
+	}
+
+	public static class Execution extends AbstractSynchronousNonBlockingStepExecution<Object> {
+		private static final long serialVersionUID = 1L;
+
+		@StepContextParameter
+		private transient TaskListener listener;
+
+		@StepContextParameter
+		private transient FilePath ws;
+
+		@Inject
+		private transient ReadYamlStep step;
+
+		@Override
+		protected Object run() throws Exception {
+			String yamlText = "";
+			if (!StringUtils.isBlank(step.getFile())) {
+				FilePath path = ws.child(step.getFile());
+				if (!path.exists()) {
+					throw new FileNotFoundException(path.getRemote() + " does not exist.");
+				}
+				if (path.isDirectory()) {
+					throw new FileNotFoundException(path.getRemote() + " is a directory.");
+				}
+				yamlText = path.readToString();
+			}
+			if (!StringUtils.isBlank(step.getText())) {
+				yamlText += "\n\n" + step.getText();
+			}
+			Iterable<Object> yaml = new Yaml().loadAll(yamlText);
```

Comment:
I tried to follow the code a bit in there, but do you know what type of objects are constructed here? Everything used in the pipeline needs to be Serializable, hence all the hacks in `readManifest` for example.
Also if there are any custom classes, they also needs to be white listed for the sandbox like in `readMavenPom`